pkg/commands/componentarchive/input/input.go

```diff
@@ -214,74 +201,110 @@ func (input *BlobInput) Read(fs vfs.FileSystem, inputFilePath string) (*BlobOutp
+	// first check if a exclude regex matches
+	for _, ex := range opts.ExcludeFiles {
+		match, err := filepath.Match(ex, path)
 		if err != nil {
-			return err
+			return false, fmt.Errorf("malformed filepath syntax %q", ex)
 		}
-		header, err := tar.FileInfoHeader(info, "")
-		if err != nil {
-			return err
+		if match {
+			return false, nil
 		}
-		relPath, err := filepath.Rel("/", path)
+	}
+
+	// if no includes are defined, include all files
+	if len(opts.IncludeFiles) == 0 {
+		return true, nil
+	}
+	// otherwise check if the file should be included
+	for _, in := range opts.IncludeFiles {
+		match, err := filepath.Match(in, path)
 		if err != nil {
-			return fmt.Errorf("unable to calculate relative path for %s: %w", path, err)
+			return false, fmt.Errorf("malformed filepath syntax %q", in)
 		}
-		// ignore the root directory.
-		if relPath == "." {
-			return nil
+		if match {
+			return true, nil
 		}
-		include, err := includeFile(relPath)
+	}
+	return false, nil
+}
+
+// TarFileSystem creates a tar archive from a filesystem.
+func TarFileSystem(ctx context.Context, fs vfs.FileSystem, root string, writer io.Writer, opts TarFileSystemOptions) error {
+	tw := tar.NewWriter(writer)
+	if opts.PreserveDir {
+		opts.root = pathutil.Base(root)
+	}
+	return addFileToTar(ctx, fs, tw, opts.root, root, opts)
+}
+
+func addFileToTar(ctx context.Context, fs vfs.FileSystem, tw *tar.Writer, path string, realPath string, opts TarFileSystemOptions) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+	log := logr.FromContextOrDiscard(ctx)
+
+	if len(path) != 0 { // do not check the root
+		include, err := opts.Included(path)
 		if err != nil {
 			return err
 		}
 		if !include {
 			return nil
 		}
+	}
+	info, err := fs.Lstat(realPath)
+	if err != nil {
+		return err
+	}
+
+	header, err := tar.FileInfoHeader(info, "")
+	if err != nil {
+		return err
+	}
+	header.Name = path
 
-		header.Name = relPath
+	switch {
+	case info.IsDir():
+		// do not write root header
+		if len(path) != 0 {
+			if err := tw.WriteHeader(header); err != nil {
+				return fmt.Errorf("unable to write header for %q: %w", path, err)
+			}
+		}
+		err := vfs.Walk(fs, realPath, func(subFilePath string, info os.FileInfo, err error) error {
+			if subFilePath == realPath {
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			relPath, err := filepath.Rel(realPath, subFilePath)
+			if err != nil {
+				return fmt.Errorf("unable to calculate relative path for %s: %w", subFilePath, err)
+			}
+			return addFileToTar(ctx, fs, tw, pathutil.Join(path, relPath), subFilePath, opts)
+		})
+		return err
+	case info.Mode().IsRegular():
 		if err := tw.WriteHeader(header); err != nil {
 			return fmt.Errorf("unable to write header for %q: %w", path, err)
 		}
-		if info.IsDir() {
-			return nil
-		}
-
-		file, err := fs.OpenFile(path, os.O_RDONLY, os.ModePerm)
+		file, err := fs.OpenFile(realPath, os.O_RDONLY, os.ModePerm)
```

Comment:
this was not changed in this PR but the file is not closed if an error occurs before in the cop function.

Reply:
done